Doppler.Sap/Services/SapService.cs

```diff
@@ -1,21 +1,130 @@
+            IHttpClientFactory httpClientFactory)
         {
             _sapTaskFactory = sapTaskFactory;
+            _sapConfig = sapConfig.Value;
+            _client = httpClientFactory.CreateClient();
         }
 
         public async Task<SapTaskResult> SendToSap(SapTask dequeuedTask)
         {
             return await _sapTaskFactory.CreateHandler(dequeuedTask);
         }
+
+        public async Task<SapBusinessPartner> TryGetBusinessPartnerByCardCode(string cardCode)
+        {
+            await StartSession();
+
+            var message = new HttpRequestMessage()
+            {
+                RequestUri = new Uri($"{_sapConfig.BaseServerUrl}BusinessPartners('{cardCode}')"),
+                Method = HttpMethod.Get
+            };
+            message.Headers.Add("Cookie", _sapCookies.B1Session);
+            message.Headers.Add("Cookie", _sapCookies.RouteId);
+
+            var sapResponse = await _client.SendAsync(message);
+
+            if (sapResponse.IsSuccessStatusCode)
+            {
+                return JsonConvert.DeserializeObject<SapBusinessPartner>((await sapResponse.Content.ReadAsStringAsync()));
+            }
+            return null;
+        }
+
+        private async Task StartSession()
+        {
+            if (_sapCookies == null || (DateTime.Now > _sessionStartedAt.AddMinutes(30)))
+            {
+                var sapResponse = await _client.SendAsync(new HttpRequestMessage
+                {
+                    RequestUri = new Uri($"{_sapConfig.BaseServerUrl}Login/"),
+                    Content = new StringContent(JsonConvert.SerializeObject(new SapConfig
+                    {
+                        CompanyDB = _sapConfig.CompanyDB,
+                        Password = _sapConfig.Password,
+                        UserName = _sapConfig.UserName
+                    })
+                        , Encoding.UTF8
+                        , "application/json"),
+                    Method = HttpMethod.Post,
+                });
+                if (sapResponse.IsSuccessStatusCode)
+                {
+                    _sapCookies = new SapLoginCookies
+                    {
+                        B1Session = sapResponse.Headers.GetValues("Set-Cookie").Where(x => x.Contains("B1SESSION")).Select(y => y.ToString().Substring(0, 46)).FirstOrDefault(),
+                        RouteId = sapResponse.Headers.GetValues("Set-Cookie").Where(x => x.Contains("ROUTEID")).Select(y => y.ToString().Substring(0, 14)).FirstOrDefault()
+                    };
+                    _sessionStartedAt = DateTime.Now;
+                    return;
+                }
+                else
+                {
+                    throw new UnauthorizedAccessException();
+                }
+            }
+            else
+            {
+                return;
+            }
+        }
+
+        public async Task<SapBusinessPartner> TryGetBusinessPartner(string cardCode, string cuit)
```

Comment:
I think this method is done in the SapTaskHandler too. In fact, I'm not sure if there should be a SapService.cs at all.

Reply:
In fact, this method was in SapTaskHandler too, and this is the correct place. We changed BusinessPartnerService in order to fix that.